File: src/main/java/com/cout970/computer2/gui/components/MonitorComponent.java
package com.cout970.computer2.gui.components;

import com.cout970.computer2.ComputerMod;
import com.cout970.computer.api.IPeripheralMonitor;
import net.darkaqua.blacksmith.api.common.gui.IGui;
import net.darkaqua.blacksmith.api.common.gui.IGuiComponent;
import net.darkaqua.blacksmith.api.common.util.ResourceReference;
import net.darkaqua.blacksmith.api.common.util.vectors.Vect2d;
import net.darkaqua.blacksmith.api.common.util.vectors.Vect2i;
import org.lwjgl.opengl.GL11;

/**
 * Created by cout970 on 07/02/2016.
 */
public class MonitorComponent implements IGuiComponent {

    private static final ResourceReference texture = new ResourceReference(ComputerMod.MOD_ID, "textures/gui/monitor_text.png");
    private IPeripheralMonitor monitor;
    private int pressedKeyNum = -1;
    private int pressedKeyCode;

    public MonitorComponent(IPeripheralMonitor monitor) {
        this.monitor = monitor;
    }

    public IPeripheralMonitor getMonitor() {
        return monitor;
    }

    @Override
    public void renderBackground(IGui gui, Vect2i mouse, float partialTicks) {
        if (pressedKeyNum != -1 && !gui.isKeyPressed(pressedKeyNum)){
            pressedKeyNum = -1;
            monitor.onKeyRelease(pressedKeyCode);
        }

        gui.getGuiRenderer().bindTexture(texture);
        GL11.glColor4f(0.0F, 1.0F, 0.0F, 1.0F);
        int cursor = monitor.getCursorPos();
        int lines = monitor.getLines();
        int columns = monitor.getColumns();
        for (int line = 0; line < lines; line++) {
            for (int column = 0; column < columns; column++) {
                int character = monitor.getChar(line * columns + column) & 255;
                if (line * columns + column == cursor && monitor.getWorldRef().getWorldTime() % 20 >= 10) {
                    character ^= 128;
                }
                if (character != 32) {
                    gui.getGuiRenderer().drawRectangleWithCustomSizedTexture(gui.getGuiStartingPoint().add(15 + column * 4, 15 + line * 4),
                            new Vect2i(4, 4), new Vect2d((character & 15) * 4, (character >> 4) * 4), new Vect2d(64, 64));
                }
            }
        }
        GL11.glColor4f(1.0F, 1.0F, 1.0F, 1.0F);
    }

    @Override
    public void renderForeground(IGui gui, Vect2i mouse) {}

    @Override
    public void onMouseClick(IGui gui, Vect2i mouse, MouseButton button) {}

    @Override
    public boolean onKeyPressed(IGui gui, int num, char key) {
        if (key == 27) return false;
        int shift = 0;
        if (gui.isShiftKeyPressed()) shift |= 64;
        if (gui.isCtrlKeyPressed()) shift |= 32;
        switch (key) {
            case 199:
                sendKey(132 | shift, num);
                break;
            case 200:
                sendKey(128 | shift, num);
                break;
            case 201:
            case 202:
            case 204:
            case 206:
            case 209:
            default:
                if (key > 0 && key <= 127) {
                    sendKey(key, num);
                }
                break;

            case 203:
                sendKey(130 | shift, num);
                break;
            case 205:
                sendKey(131 | shift, num);
                break;
            case 207:
                sendKey(133 | shift, num);
                break;
            case 208:
                sendKey(129 | shift, num);
                break;
            case 210:
                sendKey(134 | shift, num);
                break;
            case 0:
                if (num != 54 && num != 42 && num != 56 && num != 184 && num != 29 && num != 221 && num != 157 && (num < 59 || num > 70) && num != 87 && num != 88 && num != 197 && num != 183 && num != 0) {
                    sendKey(num, num);
                }
                break;
        }
        return true;
    }

    public void sendKey(int key, int num) {
        pressedKeyNum = num;
        pressedKeyCode = key;
        monitor.onKeyPressed(key);
    }
}

File: src/main/kotlin/com/cout970/computer/gui/components/MonitorComponent.kt
package com.cout970.computer.gui.components

import com.cout970.computer.api.IPeripheralMonitor
import com.cout970.computer.gui.*
import com.cout970.computer.util.resource
import com.cout970.computer.util.vector.Vec2d
import org.lwjgl.opengl.GL11

/**
 * Created by cout970 on 20/05/2016.
 */

class MonitorComponent(val monitor: IPeripheralMonitor) : IComponent {

    val TEXTURE = resource("textures/gui/monitor_text.png")
    var pressedKeyNum = -1
    var pressedKeyCode: Int = 0

    override fun drawFirstLayer(gui: IGui, mouse: Vec2d, partialTicks: Float) {
        if (pressedKeyNum != -1 && !isKeyPressed(pressedKeyNum)) {
            pressedKeyNum = -1
            monitor.onKeyRelease(pressedKeyCode)
        }

        gui.bindTexture(TEXTURE)
        GL11.glColor4f(0.0f, 1.0f, 0.0f, 1.0f)
        val cursor = monitor.cursorPos
        val lines = monitor.lines
        val columns = monitor.columns
        for (line in 0..lines - 1) {
            for (column in 0..columns - 1) {
                var character = monitor.getChar(line * columns + column) and 0xFF
                if (line * columns + column == cursor && monitor.world.worldTime % 20 >= 10) {
                    character = character xor 128
                }
                if (character != 32) {
                    gui.drawModalRectWithCustomSizedTexture(gui.getStart() + Vec2d(15 + column * 4, 15 + line * 4),
                            Vec2d(4, 4), Vec2d((character and 15) * 4, (character shr 4) * 4), Vec2d(64, 64))
                }
            }
        }
        GL11.glColor4f(1.0f, 1.0f, 1.0f, 1.0f)
    }

    override fun onKeyTyped(gui: IGui, typedChar: Char, keyCode: Int): Boolean{

        if (typedChar.toInt() == 27) return false
        var shift = 0
        if (isShiftKeyDown()) shift = shift or 64
        if (isCtrlKeyDown()) shift = shift or 32
        when (typedChar.toByte().toInt()) {
            199 -> sendKey(132 or shift, keyCode)
            200 -> sendKey(128 or shift, keyCode)
            201, 202, 204, 206,
            209 -> sendKey(typedChar.toInt(), keyCode)
            203 -> sendKey(130 or shift, keyCode)
            205 -> sendKey(131 or shift, keyCode)
            207 -> sendKey(133 or shift, keyCode)
            208 -> sendKey(129 or shift, keyCode)
            210 -> sendKey(134 or shift, keyCode)
            0 -> if (keyCode != 54 && keyCode != 42 && keyCode != 56 && keyCode != 184 && keyCode != 29 &&
                    keyCode != 221 && keyCode != 157 && (keyCode < 59 || keyCode > 70) && keyCode != 87 &&
                    keyCode != 88 && keyCode != 197 && keyCode != 183 && keyCode != 0) {
                sendKey(keyCode, keyCode)
            }
            else -> if (typedChar.toInt() > 0 && typedChar.toInt() <= 127) {
                sendKey(typedChar.toInt(), keyCode)
            }
        }
        return true
    }

    fun sendKey(key: Int, num: Int) {
        pressedKeyNum = num
        pressedKeyCode = key
        monitor.onKeyPressed(key)
    }
}